<a href="https://colab.research.google.com/github/wildcraft958/animesh_Datahack/blob/main/animesh_Datahack.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**DataHack by IIT Guwahati**

<h1> So given the time constrained i decided to use panda profiling for EDA </h1>

In [1]:
#downloading pandas profiling using pip
!pip install ydata-profiling
!pip install pydantic

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 359.5/359.5 kB 2.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.8/104.8 kB 10.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 686.1/686.1 kB 14.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 kB 30.4 MB/s eta 0:00:00
  Created wheel for htmlmin: filename=htmlmin-0.1.12-py3-none-any.whl size=27080 sha256=0988d7a3420785e1ea443ef85629e5021071f71782571a649aee8ecde63bc186
  Stored in directory: /root/.cache/pip/wheels/dd/91/29/a79cecb328d01739e64017b6fb9a1ab9d8cb1853098ec5966d
Successfully built htmlmin


<h3>Importing necessary libraries and modules</h3>

In [19]:
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score

## Data Acquisition


- Loading the datasets.


In [20]:
# Loading datasets
train_features = pd.read_csv('training_set_features.csv')
train_labels = pd.read_csv('training_set_labels.csv')
test_features = pd.read_csv('test_set_features.csv')

- Basic exploration of the data.


In [21]:
train_features.shape

(26707, 36)

In [22]:
test_features.shape

(26708, 36)

In [23]:
# Inspecting colmns
print("Train features columns:", train_features.columns)
print("Train labels columns:", train_labels.columns)
print("Test features columns:", test_features.columns)

Train features columns: Index(['respondent_id', 'xyz_concern', 'xyz_knowledge',
       'behavioral_antiviral_meds', 'behavioral_avoidance',
       'behavioral_face_mask', 'behavioral_wash_hands',
       'behavioral_large_gatherings', 'behavioral_outside_home',
       'behavioral_touch_face', 'doctor_recc_xyz', 'doctor_recc_seasonal',
       'chronic_med_condition', 'child_under_6_months', 'health_worker',
       'health_insurance', 'opinion_xyz_vacc_effective', 'opinion_xyz_risk',
       'opinion_xyz_sick_from_vacc', 'opinion_seas_vacc_effective',
       'opinion_seas_risk', 'opinion_seas_sick_from_vacc', 'age_group',
       'education', 'race', 'sex', 'income_poverty', 'marital_status',
       'rent_or_own', 'employment_status', 'hhs_geo_region', 'census_msa',
       'household_adults', 'household_children', 'employment_industry',
       'employment_occupation'],
      dtype='object')
Train labels columns: Index(['respondent_id', 'xyz_vaccine', 'seasonal_vaccine'], dtype='object')
Tes

In [24]:
train_features.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]


## Exploratory Data Analysis (EDA) using Pandas Profiling
- Importing pandas profiling.
- Generating EDA report using pandas profiling.
- Analyzing key insights and correlations.

In [18]:
from ydata_profiling import ProfileReport

#combining training data and test data in single data frame for EDA using pandas profiling
combined_data = pd.concat([train_features, test_features], axis=0)
combined_data.reset_index(drop=True, inplace=True)

# Generating pandas profiling report
profile = ProfileReport(combined_data, explorative=True)

# Saving the report to an HTML file
profile.to_file("pandas_profiling_report_Datahack.html")

/usr/local/lib/python3.10/dist-packages/ydata_profiling/profile_report.py:363: UserWarning: Try running command: 'pip install --upgrade Pillow' to avoid ValueError
  warnings.warn(


Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

## Data Preprocessing


- Handling missing values.</br>
We'll fill missing values using simple methods like filling with the most frequent value.

In [25]:
# dealing missing values using imputer
imputer = SimpleImputer(strategy='most_frequent')
train_features_imputed = pd.DataFrame(imputer.fit_transform(train_features), columns=train_features.columns)
test_features_imputed = pd.DataFrame(imputer.transform(test_features), columns=test_features.columns)


- Encoding categorical variables.
<br>We'll convert categorical variables into numerical ones using one-hot encoding.

In [26]:
# Identifying categorical columns
categorical_cols = train_features_imputed.select_dtypes(include=['object']).columns

In [27]:
# Performing one-hot encoding method
train_features_encoded = pd.get_dummies(train_features_imputed, columns=categorical_cols)
test_features_encoded = pd.get_dummies(test_features_imputed, columns=categorical_cols)

# Aligning the train and test data to ensure they have the same columns
train_features_encoded, test_features_encoded = train_features_encoded.align(test_features_encoded, join='inner', axis=1)


- Scaling numerical variables.
<br>Scaling helps models like Logistic Regression perform better by ensuring all features have the same scale.


In [28]:
# Scaling numerical variables for btter performance
scaler = StandardScaler()
train_features_scaled = pd.DataFrame(scaler.fit_transform(train_features_encoded), columns=train_features_encoded.columns)
test_features_scaled = pd.DataFrame(scaler.transform(test_features_encoded), columns=test_features_encoded.columns)

## Model Training and Evaluation


- Splitting the data for training and validation.


In [29]:
# Spliting the data for training and validation
X_train, X_val, y_train, y_val = train_test_split(train_features_scaled, train_labels, test_size=0.2, random_state=42)

- Training logistic regression models without
regularization.

In [30]:
# Training model using logistic regression without regularization
model_xyz = LogisticRegression(max_iter=1000)
model_seasonal = LogisticRegression(max_iter=1000)

In [31]:
model_xyz.fit(X_train, y_train['xyz_vaccine'])
model_seasonal.fit(X_train, y_train['seasonal_vaccine'])

LogisticRegression(max_iter=1000)

In [32]:
# Evaluating logistic regression model performance using ROC AUC scores.
val_preds_xyz = model_xyz.predict_proba(X_val)[:, 1]
val_preds_seasonal = model_seasonal.predict_proba(X_val)[:, 1]

roc_auc_xyz = roc_auc_score(y_val['xyz_vaccine'], val_preds_xyz)
roc_auc_seasonal = roc_auc_score(y_val['seasonal_vaccine'], val_preds_seasonal)

print(f'ROC AUC for xyz_vaccine: {roc_auc_xyz}')
print(f'ROC AUC for seasonal_vaccine: {roc_auc_seasonal}')
print(f'Mean ROC AUC: {(roc_auc_xyz + roc_auc_seasonal) / 2}')

ROC AUC for xyz_vaccine: 0.8333404768508013
ROC AUC for seasonal_vaccine: 0.8591162291109835
Mean ROC AUC: 0.8462283529808924


- Training logistic regression models L1 regularization.

In [33]:
# Training logistic regression model with L1 regularization for feature selection
model_xyz = LogisticRegression(max_iter=1000, penalty='l1', solver='liblinear', random_state=42)
model_seasonal = LogisticRegression(max_iter=1000, penalty='l1', solver='liblinear', random_state=42)

In [34]:
model_xyz.fit(X_train, y_train['xyz_vaccine'])
model_seasonal.fit(X_train, y_train['seasonal_vaccine'])

LogisticRegression(max_iter=1000, penalty='l1', random_state=42,
                   solver='liblinear')

In [35]:
# Evaluating logistic regression model performance with L1 regularization using ROC AUC scores.
val_preds_xyz = model_xyz.predict_proba(X_val)[:, 1]
val_preds_seasonal = model_seasonal.predict_proba(X_val)[:, 1]

roc_auc_xyz = roc_auc_score(y_val['xyz_vaccine'], val_preds_xyz)
roc_auc_seasonal = roc_auc_score(y_val['seasonal_vaccine'], val_preds_seasonal)

print(f'ROC AUC for xyz_vaccine: {roc_auc_xyz}')
print(f'ROC AUC for seasonal_vaccine: {roc_auc_seasonal}')
print(f'Mean ROC AUC: {(roc_auc_xyz + roc_auc_seasonal) / 2}')

ROC AUC for xyz_vaccine: 0.8334121221289361
ROC AUC for seasonal_vaccine: 0.8591084671530168
Mean ROC AUC: 0.8462602946409765


- Training logistic regression models L2 regularization.

In [38]:
# Training logistic regression model with L2 regularization for feature selection
model_xyz = LogisticRegression(max_iter=1000, penalty='l2', solver='liblinear', random_state=42)
model_seasonal = LogisticRegression(max_iter=1000, penalty='l2', solver='liblinear', random_state=42)

In [39]:
model_xyz.fit(X_train, y_train['xyz_vaccine'])
model_seasonal.fit(X_train, y_train['seasonal_vaccine'])

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [40]:
# Evaluating logistic regression model performance using with L2 regularization ROC AUC scores.
val_preds_xyz = model_xyz.predict_proba(X_val)[:, 1]
val_preds_seasonal = model_seasonal.predict_proba(X_val)[:, 1]

roc_auc_xyz = roc_auc_score(y_val['xyz_vaccine'], val_preds_xyz)
roc_auc_seasonal = roc_auc_score(y_val['seasonal_vaccine'], val_preds_seasonal)

print(f'ROC AUC for xyz_vaccine: {roc_auc_xyz}')
print(f'ROC AUC for seasonal_vaccine: {roc_auc_seasonal}')
print(f'Mean ROC AUC: {(roc_auc_xyz + roc_auc_seasonal) / 2}')

ROC AUC for xyz_vaccine: 0.833341947574986
ROC AUC for seasonal_vaccine: 0.8591156646049496
Mean ROC AUC: 0.8462288060899679


## Results and Analysis
- Analyzing ROC AUC scores for each vaccine prediction.
- Discussing insights from model evaluation.

1. Logistic regression model performance without regularization using ROC AUC scores.<br>
- ROC AUC for xyz_vaccine: 0.8333404768508013<br>
- ROC AUC for seasonal_vaccine: 0.8591162291109835<br>
- Mean ROC AUC: 0.8462283529808924<br>
<br>
2. Logistic regression model performance with L1 regularization using ROC AUC scores.<br>
- ROC AUC for xyz_vaccine: 0.8334121221289361<br>
- ROC AUC for seasonal_vaccine: 0.8591084671530168<br>
- Mean ROC AUC: 0.8462602946409765<br>
<br>
3. Logistic regression model performance using with L2 regularization ROC AUC scores.<br>
- ROC AUC for xyz_vaccine: 0.833341947574986<br>
- ROC AUC for seasonal_vaccine: 0.8591156646049496<br>
- Mean ROC AUC: 0.8462288060899679<br>

## Predictions on Test Set
- Making predictions on the test set using best model based on performance in ROC AUC scores (in my case it is logistic regression models L1 regularization.)
- Creating the submission file csv in required format.

In [41]:
model_xyz = LogisticRegression(max_iter=1000, penalty='l1', solver='liblinear', random_state=42)
model_seasonal = LogisticRegression(max_iter=1000, penalty='l1', solver='liblinear', random_state=42)

model_xyz.fit(X_train, y_train['xyz_vaccine'])
model_seasonal.fit(X_train, y_train['seasonal_vaccine'])

# Predictions on test set
test_preds_xyz = model_xyz.predict_proba(test_features_scaled)[:, 1]
test_preds_seasonal = model_seasonal.predict_proba(test_features_scaled)[:, 1]

##Preparing for submission in form of csv file

In [42]:
submission = pd.DataFrame({
    'respondent_id': test_features['respondent_id'],
    'xyz_vaccine': test_preds_xyz,
    'seasonal_vaccine': test_preds_seasonal
})
# Saving submission to CSV
submission.to_csv('final_submission.csv', index=False)